File: Data Structures & Algorithms/minimum-window-with-characters/submission-1.java
class Solution {
    public String minWindow(String s, String t) {
        String res = "";

        Map<Character, Integer> map = new HashMap<>();
        for(char c : t.toCharArray()){
            Integer count = map.get(c);
            if(count == null){
                count = 0;
            }
            map.put(c, count + 1);
        }

        int startIndex = 0;
        int minLength = s.length() + 1;
        int letterCount = 0;
        for(int i = 0; i < s.length(); i++){
            char curr = s.charAt(i);
            if(map.containsKey(curr)){
                int count = map.get(curr);
                count--;
                map.put(curr, count);
                if(count >= 0){
                    letterCount++;
                    if(letterCount == t.length()){
                        while(startIndex <= i && letterCount == t.length()){
                            char remove = s.charAt(startIndex);
                            if(map.containsKey(remove)){
                                int tempCount = map.get(remove);
                                if(tempCount == 0){
                                    letterCount--;
                                    if((i - startIndex) + 1 < minLength){
                                        minLength = (i - startIndex) + 1;
                                        res = s.substring(startIndex, i + 1);
                                    }
                                }
                                tempCount++;
                                map.put(remove, tempCount);
                            }
                            startIndex++;
                        }
                    }
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-1.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 0:
            return ""
        res = s[0]

        for i in range(1, len(s)):
            l,r = i-1, i+1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if len(res) < r - (l + 1):
                res = s[l+1: r]
            
            l,r = i-1, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if len(res) < r - (l + 1):
                res = s[l+1: r]

        return res